File: src/main/java/com/marsraver/WledFx/animation/BouncingBallAnimation.java
package com.marsraver.WledFx.animation;

import java.util.Random;

/**
 * Bouncing ball animation with random behavior
 */
public class BouncingBallAnimation implements LedAnimation {
    
    private static final Random random = new Random();
    private static final double GRAVITY = 0.3;
    private static final double BOUNCE_DAMPING = 0.95;
    private static final double RANDOM_CHANGE_PROBABILITY = 0.02;
    
    private int combinedWidth;
    private int combinedHeight;
    
    // Ball state
    private double ballX;
    private double ballY;
    private double velocityX;
    private double velocityY;
    private int ballRadius;
    
    // Trail effect
    private static final int MAX_TRAIL = 5;
    
    @Override
    public void init(int combinedWidth, int combinedHeight) {
        this.combinedWidth = combinedWidth;
        this.combinedHeight = combinedHeight;
        
        // Initialize ball at random position
        this.ballX = combinedWidth / 2.0 + (random.nextDouble() - 0.5) * 10;
        this.ballY = combinedHeight / 2.0 + (random.nextDouble() - 0.5) * 10;
        
        // Random initial velocity
        this.velocityX = (random.nextDouble() - 0.5) * 4;
        this.velocityY = (random.nextDouble() - 0.5) * 4;
        
        this.ballRadius = 2;
    }
    
    @Override
    public boolean update(long now) {
        // Add random velocity changes
        if (random.nextDouble() < RANDOM_CHANGE_PROBABILITY) {
            velocityX += (random.nextDouble() - 0.5) * 2;
            velocityY += (random.nextDouble() - 0.5) * 2;
        }
        
        // Update position
        ballX += velocityX;
        ballY += velocityY;
        
        // Bounce off walls
        if (ballX < ballRadius) {
            ballX = ballRadius;
            velocityX = -velocityX * BOUNCE_DAMPING;
        } else if (ballX >= combinedWidth - ballRadius) {
            ballX = combinedWidth - ballRadius - 1;
            velocityX = -velocityX * BOUNCE_DAMPING;
        }
        
        if (ballY < ballRadius) {
            ballY = ballRadius;
            velocityY = -velocityY * BOUNCE_DAMPING;
        } else if (ballY >= combinedHeight - ballRadius) {
            ballY = combinedHeight - ballRadius - 1;
            velocityY = -velocityY * BOUNCE_DAMPING;
        }
        
        // Clamp velocities to prevent excessive speed
        velocityX = Math.max(-8, Math.min(8, velocityX));
        velocityY = Math.max(-8, Math.min(8, velocityY));
        
        return true; // Keep running
    }
    
    @Override
    public int[] getPixelColor(int x, int y) {
        // Calculate distance from ball center
        double dx = x - ballX;
        double dy = y - ballY;
        double distance = Math.sqrt(dx * dx + dy * dy);
        
        if (distance <= ballRadius) {
            // Ball itself - bright color
            return new int[]{0, 255, 255}; // Cyan (to test GRB mapping)
        } else if (distance <= ballRadius + MAX_TRAIL) {
            // Trail effect
            double trailDistance = distance - ballRadius;
            double intensity = 255 * (1.0 - trailDistance / MAX_TRAIL);
            intensity = Math.max(0, Math.min(255, intensity));
            
            // Fade from cyan to black
            int r = (int)(0 * (1.0 - trailDistance / MAX_TRAIL));
            int g = (int)(255 * (1.0 - trailDistance / MAX_TRAIL));
            int b = (int)(255 * (1.0 - trailDistance / MAX_TRAIL));
            
            return new int[]{r, g, b};
        } else {
            // Off
            return new int[]{0, 0, 0};
        }
    }
    
    @Override
    public String getName() {
        return "Bouncing Ball";
    }
}

File: src/main/kotlin/com/marsraver/wledfx/animation/BouncingBallAnimation.kt
package com.marsraver.wledfx.animation

import com.marsraver.wledfx.color.RgbColor
import com.marsraver.wledfx.color.ColorUtils
import java.util.Random
import kotlin.math.sqrt

/**
 * Bouncing ball animation with random behavior.
 */
class BouncingBallAnimation : BaseAnimation() {

    private var ballX: Double = 0.0
    private var ballY: Double = 0.0
    private var velocityX: Double = 0.0
    private var velocityY: Double = 0.0
    private var ballRadius: Int = 2
    
    private val RANDOM = Random()
    private val BOUNCE_DAMPING = 0.95
    private val RANDOM_CHANGE_PROBABILITY = 0.02
    private val MAX_TRAIL = 5

    override fun getName(): String = "Bouncing Ball"
    override fun is1D(): Boolean = false
    override fun is2D(): Boolean = true

    override fun onInit() {
        ballX = width / 2.0 + (RANDOM.nextDouble() - 0.5) * 10
        ballY = height / 2.0 + (RANDOM.nextDouble() - 0.5) * 10
        velocityX = (RANDOM.nextDouble() - 0.5) * 4
        velocityY = (RANDOM.nextDouble() - 0.5) * 4
        ballRadius = 1
    }

    override fun update(now: Long): Boolean {
        if (RANDOM.nextDouble() < RANDOM_CHANGE_PROBABILITY) {
            velocityX += (RANDOM.nextDouble() - 0.5) * 2
            velocityY += (RANDOM.nextDouble() - 0.5) * 2
        }

        ballX += velocityX
        ballY += velocityY

        if (ballX < ballRadius) {
            ballX = ballRadius.toDouble()
            velocityX = -velocityX * BOUNCE_DAMPING
        } else if (ballX >= width - ballRadius) {
            ballX = (width - ballRadius - 1).toDouble()
            velocityX = -velocityX * BOUNCE_DAMPING
        }

        if (ballY < ballRadius) {
            ballY = ballRadius.toDouble()
            velocityY = -velocityY * BOUNCE_DAMPING
        } else if (ballY >= height - ballRadius) {
            ballY = (height - ballRadius - 1).toDouble()
            velocityY = -velocityY * BOUNCE_DAMPING
        }
        
        // Speed check?
        // Original logic didn't use speed param.
        // We can use paramSpeed to scale velocity if we want, but let's stick to original behavior for now
        // Or constrain logic.

        velocityX = velocityX.coerceIn(-8.0, 8.0)
        velocityY = velocityY.coerceIn(-8.0, 8.0)

        return true
    }

    override fun getPixelColor(x: Int, y: Int): RgbColor {
        val dx = x - ballX
        val dy = y - ballY
        val distance = sqrt(dx * dx + dy * dy)

        return when {
            distance <= ballRadius -> paramColor
            distance <= ballRadius + MAX_TRAIL -> {
                val trailDistance = distance - ballRadius
                val factor = (1.0 - trailDistance / MAX_TRAIL).coerceIn(0.0, 1.0)
                ColorUtils.scaleBrightness(paramColor, factor)
            }
            else -> RgbColor.BLACK
        }
    }
}
